### PR TITLE
NAT Traversal: improved connection consistency and fixed port dropping

### PR DESCRIPTION
**Pull request by sbsends**

### CODE REVIEW

File: network/nat/plugin.go [hunk 1]
```diff
@@ -86,18 +86,28 @@ func (p *plugin) Startup(n *network.Network) {
...
+
 	// Set peer information based off of port mapping info.
 	n.Address = info.String()
 	n.ID = peer.CreateID(n.Address, n.GetKeys().PublicKey)
 
-	log.Info().Msgf("other peers may connect to you through the address %s.", n.Address)
+	go func() {
+		for {
+			time.Sleep(60 * time.Second)
+			_, err = gateway.AddPortMapping("tcp", p.internalPort, "noise", 60*time.Second)
```
Comment (jack0) [inline network/nat/plugin.go]: Thanks for your contribution. When have you noticed the port dropping? Is it replicable?

**Comment by sbsends** [comment]
While trying to use NAT traversal I noticed that my router was correctly routing ports (upnpc -l) but nmap showed that the listening port was closed on the host. 

The listener is passed the peer address but subsequently begins listening on the port number being forwarded, not the actual target.

File: network/nat/plugin.go (hunk 1 above)
Comment (sbsends) [reply network/nat/plugin.go]: When connecting via my router, connections were refused ~50% of the time with a 1s timeout. We should quantifiable measure whether we need to ping the router but qualitatively I had a smoother experience with connecting/disconnecting.

Comment (sbsends) [reply network/nat/plugin.go]: If you are referring to the local port, it is not being correctly listened to... Use upnpc and nmap to verify

Comment (sbsends) [reply network/nat/plugin.go]: @jack0, I pushed a new commit that removes this. Please let the NAT traversal fix through.

**Comment by iwasaki-kenta** [comment]
> While trying to use NAT traversal I noticed that my router was correctly routing ports (upnpc -l) but nmap showed that the listening port was closed on the host.
> 
> The listener is passed the peer address but subsequently begins listening on the port number being forwarded, not the actual target.

Noise v2 fixes this issue by introducing a new `ExternalPort` option which is exposed under a nodes identity. A nodes `InternalPort` is what `net.Listener` begins listening on :smile:.

If the issue still exists in some shape or form however, don't hesitate to comment. I'll close the PR for the time being as Noise v2 introduces a lot of breaking changes.

**Pull request closed without merging**